File: MathCore.DSP/Filters/ButterworthBandStop.cs
﻿using System.Diagnostics;

using static System.Math;
using static MathCore.Polynom.Array;
// ReSharper disable InconsistentNaming

namespace MathCore.DSP.Filters;

public class ButterworthBandStop : ButterworthFilter
{
    private static void CheckFrequencies(double dt, double fpl, double fsl, double fsh, double fph)
    {
        if (dt <= 0)
            throw new InvalidOperationException($"Период дискретизации dt={dt} не может быть меньше, либо равен нулю")
            {
                Data =
                {
                    { nameof(dt), dt },
                    { "fd", 1/dt },
                    { nameof(fpl), fpl },
                    { nameof(fsl), fsl },
                    { nameof(fsh), fsh },
                    { nameof(fph), fph },
                }
            };

        if (1 / dt == 0)
            throw new InvalidOperationException("Частота дискретизации не может быть равна нулю")
            {
                Data =
                {
                    { nameof(dt), dt },
                    { "fd", 1/dt },
                    { nameof(fpl), fpl },
                    { nameof(fsl), fsl },
                    { nameof(fsh), fsh },
                    { nameof(fph), fph },
                }
            };

        if (fpl >= fsl)
            throw new InvalidOperationException($"Нижняя частота пропускания fpl должна быть ниже нижней частоты среза fsl\r\n  dt={dt}\r\n  fpl={fpl}\r\n  fsl={fsl}\r\n  fsh={fsh}\r\n  fph={fsh}")
            {
                Data =
                {
                    { nameof(dt), dt },
                    { "fd", 1/dt },
                    { nameof(fpl), fpl },
                    { nameof(fsl), fsl },
                    { nameof(fsh), fsh },
                    { nameof(fph), fph },
                }
            };

        if (fsl >= fsh)
            throw new InvalidOperationException($"Нижняя частота среза fsl должна быть ниже верхней частоты среза fsh\r\n  dt={dt}\r\n  fpl={fpl}\r\n  fsl={fsl}\r\n  fsh={fsh}\r\n  fph={fsh}")
            {
                Data =
                {
                    { nameof(dt), dt },
                    { "fd", 1/dt },
                    { nameof(fpl), fpl },
                    { nameof(fsl), fsl },
                    { nameof(fsh), fsh },
                    { nameof(fph), fph },
                }
            };

        if (fsh >= fph)
            throw new InvalidOperationException($"Верхняя частота среза fsh должна быть ниже верхней частоты пропускания fph\r\n  dt={dt}\r\n  fpl={fpl}\r\n  fsl={fsl}\r\n  fsh={fsh}\r\n  fph={fsh}")
            {
                Data =
                {
                    { nameof(dt), dt },
                    { "fd", 1/dt },
                    { nameof(fpl), fpl },
                    { nameof(fsl), fsl },
                    { nameof(fsh), fsh },
                    { nameof(fph), fph },
                }
            };

        if (fph >= 1 / dt / 2)
            throw new InvalidOperationException($"Верхняя частота пропускания fph должна быть ниже половины частоты дискретизации fd={1 / dt} (1 / (dt={dt}))\r\n  dt={dt}\r\n  fpl={fpl}\r\n  fsl={fsl}\r\n  fsh={fsh}\r\n  fph={fsh}")
            {
                Data =
                {
                    { nameof(dt), dt },
                    { "fd", 1/dt },
                    { nameof(fpl), fpl },
                    { nameof(fsl), fsl },
                    { nameof(fsh), fsh },
                    { nameof(fph), fph },
                }
            };
    }

    /// <summary>Формирование спецификации фильтра</summary>
    /// <param name="dt">Период дискретизации сигнала</param>
    /// <param name="fpl">Частота нижней границы полосы пропускания</param>
    /// <param name="fsl">Частота нижней границы полосы заграждения</param>
    /// <param name="fsh">Частота верхней границы полосы заграждения</param>
    /// <param name="fph">Частота верхней границы полосы пропускания</param>
    /// <param name="Gp">Уровень АЧХ в полосе пропускания (в единицах 0..1)</param>
    /// <param name="Gs">Неравномерность АЧХ в полосе заграждения (в единицах 0..1)</param>
    /// <returns>Спецификация фильтра</returns>
    private static Specification GetSpecification(
        double dt,
        double fpl,
        double fsl,
        double fsh,
        double fph,
        double Gp = 0.891250938,
        double Gs = 0.01)
    {
        CheckFrequencies(dt, fpl, fsl, fsh, fph);

        if (Gp <= Gs)
            throw new ArgumentOutOfRangeException(
                nameof(Gp), Gp, $"Уровень АЧХ в полосе пропускания Gp={Gp} был меньше, либо равен уровню АЧХ в полосе заграждения Gs={Gs}");

        var Fpl = ToDigitalFrequency(fpl, dt);
        var Fsl = ToDigitalFrequency(fsl, dt);
        var Fsh = ToDigitalFrequency(fsh, dt);
        var Fph = ToDigitalFrequency(fph, dt);

        var Wpl = Consts.pi2 * Fpl;
        var Wsl = Consts.pi2 * Fsl;
        var Wsh = Consts.pi2 * Fsh;
        var Wph = Consts.pi2 * Fph;

        var Wc = Wsl * Wsh;
        var dW = Wsh - Wsl;

        // Выбор опорной частоты
        // Если   Wc / Wph > Wpl
        // то есть      Wc > Wpl*Wph
        // то есть Wsl*Wsh > Wpl*Wph
        // то есть центральная частота по границам подавления > центральной частоты по границам пропускания
        // то выбираем в качестве опорной частоты выбираем верхнюю границу пропускания
        // иначе, выбираем нижнюю границу пропускания
        var Ws = Wc / Wph > Wpl
            ? Wph
            : Wpl;
        var Fp = Abs(dW * Ws / (Wc - Ws.Pow2())) / Consts.pi2;
        const double Fs = 1 / Consts.pi2;

        // Для передачи информации о граничных частотах в спецификацию аналогвого прототипа перечситываем частоты цифрового фильтра обратно
        var fp = ToAnalogFrequency(Fp, dt);
        var fs = ToAnalogFrequency(Fs, dt);

        return new(dt, fp, fs, Gp, Gs);
    }

    /// <summary>Расчёт коэффициентов полиномов числителя из знаменателя передаточной функции фильтра</summary>
    /// <param name="fpl">Нижняя частота полосы пропускания</param>
    /// <param name="fsl">Нижняя частота полосы подавления</param>
    /// <param name="fsh">Верхняя частота полосы подавления</param>
    /// <param name="fph">Верхняя частота полосы пропускания</param>
    /// <param name="Spec">Спецификация фильтра</param>
    /// <returns>Кортеж, содержащий массивы A - коэффициенты полинома знаменателя и B - коэффициенты полинома числителя</returns>
    private static (double[] A, double[] B) Initialize(double fpl, double fsl, double fsh, double fph, Specification Spec)
    {
        // Пересчитываем аналоговые частоты полосы заграждения в цифровые
        var dt = Spec.dt;
        CheckFrequencies(dt, fpl, fsl, fsh, fph);

        var Wpl = Consts.pi2 * ToDigitalFrequency(fpl, dt);
        var Wsl = Consts.pi2 * ToDigitalFrequency(fsl, dt);
        var Wsh = Consts.pi2 * ToDigitalFrequency(fsh, dt);
        var Wph = Consts.pi2 * ToDigitalFrequency(fph, dt);

        var Wc = Wsl * Wsh;
        var dW = Wsh - Wsl;
        var sqrtWc = Wc.Sqrt();
        var Wp = Wc / Wph > Wpl
            ? Wph
            : Wpl;
        var W0 = Abs(dW * Wp / (Wc - Wp.Pow2()));

        var N = (int)Ceiling(Log(Spec.kEps) / Log(Spec.kW));
        Debug.Assert(N > 0, $"N > 0 :: {N} > 0");
        var poles = GetNormPoles(N, Spec.EpsP, W0);

        // Переносим нули и полюса аналогового нормированного ФНЧ в полосы ПЗФ
        var pzf_zeros = Enumerable.Range(0, 2 * N).Select(i => i % 2 == 0 ? Complex.ImValue(sqrtWc) : Complex.ImValue(-sqrtWc));
        var pzf_poles = TransformToBandStopW(poles, Wsl, Wsh);

        // Преобразуем аналоговые нули и полюса в нули и полюса цифрового фильтра с помощью Z-преобразования
        var z_zeros = ToZArray(pzf_zeros, dt);
        var z_poles = ToZArray(pzf_poles, dt);

        // Вычисляем коэффициент нормировки фильтра на нулевой частоте 
        var g_norm = (N.IsOdd() ? 1 : Spec.Gp)
            / (z_zeros.Multiply(z => 1 - z) / z_poles.Multiply(z => 1 - z)).Abs;

        // Определяем массивы нулей коэффициентов полиномов знаменателя и числителя
        var B = GetCoefficientsInverted(z_zeros).ToArray(b => b.Re * g_norm);
        var A = GetCoefficientsInverted(z_poles).ToRe();

        return (A, B);
    }

    /// <summary>Инициализация нового эллиптического полосозаграждающего фильтра (ПЗФ)</summary>
    /// <param name="dt">Период дискретизации цифрового сигнала</param>
    /// <param name="fpl">Нижняя граница полосы пропускания</param>
    /// <param name="fsl">Нижняя граница полосы подавления</param>
    /// <param name="fsh">Верхняя граница полосы подавления</param>
    /// <param name="fph">Верхняя граница полосы пропускания</param>
    /// <param name="Gp">Уровень АЧХ в полосе пропускания (по умолчанию -1дБ)</param>
    /// <param name="Gs">Уровень АЧХ в полосе подавления (по умолчанию -30дБ)</param>
    public ButterworthBandStop(
        double dt,
        double fpl,
        double fsl,
        double fsh,
        double fph,
        double Gp = 0.89125093813374556,
        double Gs = 0.01) 
        : this(fpl, fsl, fsh, fph, GetSpecification(dt, fpl, fsl, fsh, fph, Gp, Gs)) { }

    /// <summary>Инициализация нового эллиптического полосозаграждающего фильтра (ПЗФ)</summary>
    /// <param name="fpl">Нижняя граница полосы пропускания</param>
    /// <param name="fsl">Нижняя граница полосы подавления</param>
    /// <param name="fsh">Верхняя граница полосы подавления</param>
    /// <param name="fph">Верхняя граница полосы пропускания</param>
    /// <param name="Spec">Спецификация фильтра</param>
    private ButterworthBandStop(double fpl,double fsl, double fsh, double fph, Specification Spec) 
        : this(Initialize(fpl, fsl, fsh, fph, Spec), Spec) { }

    /// <summary>Инициализация нового эллиптического полосозаграждающего фильтра (ПЗФ)</summary>
    /// <param name="Polynoms">Кортеж с коэффициентами полиномов знаменателя и числителя функции фильтра</param>
    /// <param name="Spec">Спецификация фильтра</param>
    private ButterworthBandStop((double[] A, double[] B) Polynoms, Specification Spec) : this(Polynoms.B, Polynoms.A, Spec) { }

    /// <summary>Инициализация нового эллиптического полосозаграждающего фильтра (ПЗФ)</summary>
    /// <param name="B">Коэффициенты полинома числителя</param>
    /// <param name="A">Коэффициенты полинома знаменателя</param>
    /// <param name="Spec">Спецификация фильтра</param>
    private ButterworthBandStop(double[] B, double[] A, Specification Spec) : base(B, A, Spec) { }
}

File: MathCore.DSP/Filters/Python/filters/butterworth_bandstop.py
# Полосозаграждающий фильтр Баттерворта (порт C# ButterworthBandStop)
import numpy as np
from math import ceil, log, pi, fabs

from .butterworth_filter import ButterworthFilter
from .analog_based_filter import AnalogBasedFilter, Specification
from .digital_filter import DigitalFilter

pi2 = 2.0 * pi


class ButterworthBandStop(ButterworthFilter):
    """
    Цифровой полосозаграждающий фильтр Баттерворта (ПЗФ).

    Нули в аналоговой области: 2N штук на мнимой оси в точке ±j·√Wc,
    что соответствует центру полосы заграждения.
    """

    @staticmethod
    def _get_specification(dt: float, fpl: float, fsl: float, fsh: float, fph: float,
                           Gp: float, Gs: float) -> Specification:
        """Формирует спецификацию ПЗФ через эквивалентный ФНЧ-прототип."""
        Fpl = DigitalFilter.to_digital_frequency(fpl, dt)
        Fsl = DigitalFilter.to_digital_frequency(fsl, dt)
        Fsh = DigitalFilter.to_digital_frequency(fsh, dt)
        Fph = DigitalFilter.to_digital_frequency(fph, dt)

        Wpl, Wsl = pi2 * Fpl, pi2 * Fsl
        Wsh, Wph = pi2 * Fsh, pi2 * Fph

        Wc = Wsl * Wsh   # квадрат центральной частоты полосы заграждения
        dW = Wsh - Wsl   # ширина полосы заграждения

        # опорная частота: верхняя или нижняя граница пропускания
        Wp = Wph if (Wc / Wph > Wpl) else Wpl
        Fp = fabs(dW * Wp / (Wc - Wp ** 2)) / pi2
        Fs = 1.0 / pi2

        fp = DigitalFilter.to_analog_frequency(Fp, dt)
        fs = DigitalFilter.to_analog_frequency(Fs, dt)
        return Specification(dt, fp, fs, Gp, Gs)

    @staticmethod
    def _initialize(fpl: float, fsl: float, fsh: float, fph: float, spec: Specification):
        """
        Расчёт коэффициентов ПЗФ Баттерворта.

        Алгоритм
        --------
        1. Вычисляем опорный коэффициент W0 и порядок N.
        2. Нормированные полюса ФНЧ-прототипа с масштабом W0.
        3. Нули в аналоговой области: 2N штук ±j·√Wc.
        4. LP→BS трансформация полюсов.
        5. Билинейное z-преобразование нулей и полюсов.
        6. Нормировка на нулевой частоте (z=1).
        """
        dt = spec.dt
        Wpl = pi2 * DigitalFilter.to_digital_frequency(fpl, dt)
        Wsl = pi2 * DigitalFilter.to_digital_frequency(fsl, dt)
        Wsh = pi2 * DigitalFilter.to_digital_frequency(fsh, dt)
        Wph = pi2 * DigitalFilter.to_digital_frequency(fph, dt)

        Wc = Wsl * Wsh
        dW = Wsh - Wsl
        sqrt_wc = np.sqrt(Wc)

        # опорная частота и масштаб для ФНЧ-прототипа
        Wp = Wph if (Wc / Wph > Wpl) else Wpl
        W0 = fabs(dW * Wp / (Wc - Wp ** 2))

        N = int(ceil(log(spec.kEps) / log(spec.kW)))

        poles = ButterworthFilter.get_norm_poles(N, spec.EpsP, W0)

        # нули в аналоговой области: N пар ±j·√Wc
        pzf_zeros = [complex(0.0, sqrt_wc if i % 2 == 0 else -sqrt_wc)
                     for i in range(2 * N)]

        # LP→BS трансформация полюсов
        pzf_poles = AnalogBasedFilter.transform_to_band_stop_w(poles, Wsl, Wsh)

        # билинейное z-преобразование
        z_zeros = DigitalFilter.to_z_array(pzf_zeros, dt)
        z_poles = DigitalFilter.to_z_array(pzf_poles, dt)

        # нормировка: усиление на нулевой частоте (z=1)
        gain_zeros = np.prod([1.0 - z for z in z_zeros])
        gain_poles = np.prod([1.0 - z for z in z_poles])
        k0 = 1.0 if N % 2 != 0 else spec.Gp
        g_norm = k0 / abs(gain_zeros / gain_poles)

        B = np.poly(z_zeros).real * g_norm
        A = np.poly(z_poles).real

        return A, B

    def __init__(self, dt: float, fpl: float, fsl: float, fsh: float, fph: float,
                 Gp: float = 0.891250938, Gs: float = 0.01):
        """
        Создаёт ПЗФ Баттерворта.

        Параметры
        ---------
        dt  : период дискретизации (с)
        fpl : нижняя граница полосы пропускания (Гц)
        fsl : нижняя граница полосы заграждения (Гц)
        fsh : верхняя граница полосы заграждения (Гц)
        fph : верхняя граница полосы пропускания (Гц)
        Gp  : коэффициент передачи на границе пропускания
        Gs  : коэффициент передачи на границе заграждения
        """
        spec = ButterworthBandStop._get_specification(dt, fpl, fsl, fsh, fph, Gp, Gs)
        A, B = ButterworthBandStop._initialize(fpl, fsl, fsh, fph, spec)
        super().__init__(B, A, spec)
